### PR TITLE
Add datatypes for Sybila tools

### PR DESCRIPTION
**Pull request by xtrojak**
This PR adds specific datatypes for tools and visualisations developed in [Systems biology laboratory](https://sybila.fi.muni.cz/tools.html) - eBCSgen ([tools](https://github.com/sybila/eBCSgen), [vis](https://github.com/sybila/eBCSgen/tree/master/Galaxy/Visualisations)), Pithya ([tool](https://github.com/sybila/pithya-core), [vis](https://github.com/sybila/pithya_js)), and Parasim ([tool](https://github.com/sybila/parasim), [vis](https://github.com/sybila/robustness_js)). These are currently hosted only locally on our ancient [Galaxy instance](https://biodivine-vm.fi.muni.cz/galaxy/), and this is a preparation step before contributing them to the Galaxy toolshed (resp. to the core in the case of visualisations).  Seven `Text`-based and two `Json`-based datatypes are introduced, with respective sniffers and `set_peek` where applicable.

CoFest 2022 contribution.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:

I added automated tests for introduced datatypes, they cover `sniff` and `set_peek` where applicable. I tried inserting the sniffers in  `datatypes_conf.xml.sample` config to the correct position so that no other sniffers are broken.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Review by bernt-matthias — COMMENTED** [review]
Tests would be great. Maybe like https://github.com/galaxyproject/galaxy/blob/dev/test/unit/data/datatypes/test_qiime2.py or inline https://github.com/galaxyproject/galaxy/blob/3ea593d67d755664be9a680e557f3e8078a6d7d6/lib/galaxy/datatypes/text.py#L61

**Comment by xtrojak** [comment]
> Tests would be great. Maybe like https://github.com/galaxyproject/galaxy/blob/dev/test/unit/data/datatypes/test_qiime2.py or inline
> 
> https://github.com/galaxyproject/galaxy/blob/3ea593d67d755664be9a680e557f3e8078a6d7d6/lib/galaxy/datatypes/text.py#L61

@bernt-matthias I introduced automated tests for the introduced datatypes in [5241c60](https://github.com/galaxyproject/galaxy/pull/14362/commits/5241c609a1fdb38ea89ce5d94582d6fdc2445d5a). They cover `sniff` and `set_peek` where applicable.

**Comment by xtrojak** [comment]
Can anyone provide an assistance, please? I'm unable to replicate the intergration test errors.

**Comment by bernt-matthias** [comment]
Me neither. But maybe fixing the linter errors is a good 1st step: 

```
 test/unit/data/datatypes/test_storm.py:37: error: "MockDataset" has no
attribute "peek"  [attr-defined]
            assert dataset.peek == expected_peek
                   ^
test/unit/data/datatypes/test_bcsl.py:36: error: "MockDataset" has no attribute
"peek"  [attr-defined]
            assert dataset.peek == expected_peek
```

**Comment by xtrojak** [comment]
> Me neither. But maybe fixing the linter errors is a good 1st step:
> 
> ```
>  test/unit/data/datatypes/test_storm.py:37: error: "MockDataset" has no
> attribute "peek"  [attr-defined]
>             assert dataset.peek == expected_peek
>                    ^
> test/unit/data/datatypes/test_bcsl.py:36: error: "MockDataset" has no attribute
> "peek"  [attr-defined]
>             assert dataset.peek == expected_peek
> ```

The issue is `MockDataset` does not have the attribute `peek` defined (well, it's just a mock object); I added the respective test files to a `mypy` whitelist (like many other tests did before).

**Comment by mvdbeek** [comment]
@xtrojak  You'll want to end your test data with newlines
https://github.com/galaxyproject/galaxy/pull/14362/files#diff-869bcd953e29cd2f56ce36e6d19c482ec9eec9389fb494fa24a3f08cf21ef0d0, https://github.com/galaxyproject/galaxy/pull/14362/files#diff-3ec34efd75c8bcfd38bc36f57ebf103d37a5a676f0fc5adca6bdd381b81bbced and https://github.com/galaxyproject/galaxy/pull/14362/files#diff-e4c4f430a83875b63977ce8829a74437def48a2b54214877746ed15d60ae03cd

You can run individual tests with `./run_test.sh -integration test/integration -- -k <name_of_failing_test>`

**Comment by xtrojak** [comment]
Thank you @mvdbeek !

File: lib/galaxy/datatypes/text.py [hunk 1]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+
+    file_ext = "bcsl.model"
+
+    def sniff_prefix(self, file_prefix: FilePrefix):
+        """
+        Determines whether the file is in .bcsl.model format
+        """
+        content = open(file_prefix.filename, "r").read()
+        keywords = ["#! rules", "#! inits", "#! definitions"]
+        return any(keyword in content for keyword in keywords)
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
        keywords = ["#! rules", "#! inits", "#! definitions"]
        return any(keyword in file_prefix.contents_header for keyword in keywords)
```

Comment (mvdbeek) [reply lib/galaxy/datatypes/text.py]: Is there a specific order in which the keywords have to appear ? Do they need to appear at the start of a line ? Anything to make this more specific would be much appreciated.

File: lib/galaxy/datatypes/text.py [hunk 2]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+        else:
+            dataset.peek = "file does not exist"
+            dataset.blurb = "file purged from disk"
+
+    def _looks_like_bcsl_ts(self, file_prefix: FilePrefix):
+        content = open(file_prefix.filename, "r").read()
+        keywords = ['"edges":', '"nodes":', '"ordering":', '"initial":']
+        if all(keyword in content for keyword in keywords):
+            try:
+                json.load(open(file_prefix.filename, "r"))
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]: We definitely can't load the entire file into memory here.

File: lib/galaxy/datatypes/text.py [hunk 3]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+
+    file_ext = "storm.sample"
+
+    def sniff_prefix(self, file_prefix: FilePrefix):
+        """
+        Determines whether the file is in .storm.sample format
+        """
+        content = open(file_prefix.filename, "r").read()
+        keywords = ["Storm-pars", "Result (initial states)"]
+        return all(keyword in content for keyword in keywords)
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
        keywords = ["Storm-pars", "Result (initial states)"]
        return all(keyword in file_prefix.content_headers for keyword in keywords)
```

File: lib/galaxy/datatypes/text.py [hunk 4]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+
+    file_ext = "storm.check"
+
+    def sniff_prefix(self, file_prefix: FilePrefix):
+        """
+        Determines whether the file is in .storm.check format
+        """
+        content = open(file_prefix.filename, "r").read()
+        keywords = ["Storm ", "Result (for initial states)"]
+        return all(keyword in content for keyword in keywords)
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
        keywords = ["Storm ", "Result (for initial states)"]
        return all(keyword in file_prefix.content_headers for keyword in keywords)
```

File: lib/galaxy/datatypes/text.py [hunk 5]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+
+    def set_peek(self, dataset):
+        if not dataset.dataset.purged:
+            result = open(dataset.file_name, "r")
+            answer = ""
+            for line in result.readlines():
+                if "Result (for initial states):" in line:
+                    answer = line.split()[-1]
+            dataset.peek = "Model checking result: {}".format(answer)
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
            with open(dataset.file_name, "r") as result:
                answer = ""
                for line in result:
                    if "Result (for initial states):" in line:
                        answer = line.split()[-1]
                        break
            dataset.peek = "Model checking result: {}".format(answer)
```

File: lib/galaxy/datatypes/text.py [hunk 6]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+
+    file_ext = "ctl.result"
+
+    def sniff_prefix(self, file_prefix: FilePrefix):
+        """
+        Determines whether the file is in .ctl.result format
+        """
+        content = open(file_prefix.filename, "r").read()
+        keywords = ["Result:", "Number of satisfying states:"]
+        return all(keyword in content for keyword in keywords)
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
        keywords = ["Result:", "Number of satisfying states:"]
        return all(keyword in file_prefix.content_headers for keyword in keywords)
```

File: lib/galaxy/datatypes/text.py [hunk 7]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+
+    def set_peek(self, dataset):
+        if not dataset.dataset.purged:
+            result = open(dataset.file_name, "r")
+            answer = ""
+            for line in result.readlines():
+                if "Result:" in line:
+                    answer = line.split()[-1]
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
            with open(dataset.file_name, "r") as result:
                answer = ""
                for line in result:
                    if "Result:" in line:
                        answer = line.split()[-1]
```

File: lib/galaxy/datatypes/text.py [hunk 8]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+    file_ext = "pithya.property"
+
+    def sniff_prefix(self, file_prefix: FilePrefix):
+        """
+        Determines whether the file is in .pithya.property format
+        """
+        content = open(file_prefix.filename, "r").read()
+        if re.search(r":\?[a-zA-Z0-9_]+[ ]*=", content) is not None:
+            return True
+        return False
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
        return re.search(r":\?[a-zA-Z0-9_]+[ ]*=", file_prefix.content_headers) is not None
```

File: lib/galaxy/datatypes/text.py [hunk 9]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+
+    def sniff_prefix(self, file_prefix: FilePrefix):
+        """
+        Determines whether the file is in .pithya.model format
+        """
+        content = open(file_prefix.filename, "r").read()
+        keywords = ["VARS", "EQ", "THRES"]
+        if all(keyword in content for keyword in keywords):
+            return True
+        return False
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
        keywords = ["VARS", "EQ", "THRES"]
        return all(keyword in file_prefix.header_contents for keyword in keywords)
```

File: lib/galaxy/datatypes/text.py [hunk 10]
```diff
@@ -1093,6 +1093,203 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+        if self._looks_like_json(file_prefix):
+            is_pithya_result = self._looks_like_pithya_result(file_prefix)
+        return is_pithya_result
+
+    def _looks_like_pithya_result(self, file_prefix: FilePrefix):
+        content = open(file_prefix.filename, "r").read()
+        keywords = ['"variables":', '"states":', '"parameter_values":', '"results":']
+        if all(keyword in content for keyword in keywords):
+            try:
+                json.load(open(file_prefix.filename, "r"))
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]: Again, we can't load up the entire file into memory

File: lib/galaxy/datatypes/text.py (hunk 1 above)
Comment (xtrojak) [reply lib/galaxy/datatypes/text.py]: There is no strict order given, they kind of identify individual sections of the model, but these can be in arbitrary order.
But you are right, they should appear at the start of a line, I changed this in [769bbcd](https://github.com/galaxyproject/galaxy/pull/14362/commits/769bbcdbfda98f506f69f2efa76537d4b85ab12b).

File: lib/galaxy/datatypes/text.py (hunk 2 above)
Comment (xtrojak) [reply lib/galaxy/datatypes/text.py]: The same approach is used by the base `Json` class, but yeah, it employs some additional checks for file size, so I reused its functionality in [6e6d7d6](https://github.com/galaxyproject/galaxy/pull/14362/commits/6e6d7d6ccb6ff6448ed0956962669c7bdb069c6a).

**Comment by xtrojak** [comment]
Thank you for the review @mvdbeek, let me know if some additional changes are needed.

**Review by mvdbeek — APPROVED** [review]
Awesome, thank you @xtrojak!

**Pull request merged**